types/src/transaction/mod.rs

```diff
@@ -538,6 +762,158 @@ impl TransactionPayload {
+        match self {
+            TransactionPayload::EntryFunction(entry_function) => {
+                Ok(TransactionExecutableRef::EntryFunction(entry_function))
+            },
+            TransactionPayload::Script(script) => Ok(TransactionExecutableRef::Script(script)),
+            TransactionPayload::Multisig(multisig) => Ok(multisig.as_transaction_executable_ref()),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { executable, .. }) => {
+                Ok(executable.as_ref())
+            },
+            TransactionPayload::ModuleBundle(_) => {
+                Err(format_err!("ModuleBundle variant is deprecated"))
+            },
+        }
+    }
+
+    pub fn extra_config(&self) -> TransactionExtraConfig {
+        match self {
+            TransactionPayload::Script(_)
+            | TransactionPayload::EntryFunction(_)
+            | TransactionPayload::ModuleBundle(_) => TransactionExtraConfig::V1 {
+                multisig_address: None,
+                replay_protection_nonce: None,
+            },
+            TransactionPayload::Multisig(multisig) => TransactionExtraConfig::V1 {
+                multisig_address: Some(multisig.multisig_address),
+                replay_protection_nonce: None,
+            },
+            TransactionPayload::Payload(TransactionPayloadInner::V1 { extra_config, .. }) => {
+                extra_config.clone()
+            },
+        }
+    }
+
+    // Used in transaction emitter when printing errors.
+    pub fn payload_type(&self) -> Cow<'static, str> {
+        if let TransactionPayload::ModuleBundle(_) = &self {
+            return "module_bundle".into();
+        }
+        match self {
+            TransactionPayload::Script(_) => "script".into(),
+            TransactionPayload::ModuleBundle(_) => "module_bundle".into(),
+            TransactionPayload::EntryFunction(entry_function) => format!(
+                "entry {}::{}",
+                entry_function.module(),
+                entry_function.function()
+            )
+            .into(),
+            TransactionPayload::Multisig(_) => "multisig".into(),
+            TransactionPayload::Payload(TransactionPayloadInner::V1 {
+                executable,
+                extra_config: _,
+            }) => match executable {
+                TransactionExecutable::EntryFunction(entry_function) => format!(
+                    "inner_entry {}::{}",
+                    entry_function.module(),
+                    entry_function.function()
+                )
+                .into(),
+                TransactionExecutable::Script(_) => "inner_script".into(),
+                TransactionExecutable::Empty => "inner_empty".into(),
+            },
+        }
+    }
+
+    // Used in sdk and a lot of tests when upgrading current payload format to the new format.
+    pub fn upgrade_payload(
+        self,
+        use_txn_payload_v2_format: bool,
+        use_orderless_transactions: bool,
+    ) -> Self {
```

Comment:
I'm assuming this can only happen before signing the txn?

Reply:
Yes. Can happen only before signing the transaction.